osquery/tables/system/darwin/crash_logs.cpp

```diff
@@ -0,0 +1,154 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#include <boost/algorithm/string/predicate.hpp>
+#include <boost/algorithm/string/split.hpp>
+#include <boost/algorithm/string/trim.hpp>
+#include <boost/format.hpp>
+#include <boost/regex.hpp>
+
+#include <osquery/filesystem.h>
+#include <osquery/logger.h>
+#include <osquery/tables.h>
+
+#include "osquery/core/conversions.h"
+#include "osquery/tables/system/system_utils.h"
+
+namespace fs = boost::filesystem;
+namespace alg = boost::algorithm;
+
+namespace osquery {
+namespace tables {
+
+/// Locationg of the system application crash logs in OS X
+const std::string kDiagnosticReportsPath = "/Library/Logs/DiagnosticReports";
+/// Map of the values we currently parse out of the log file
+const std::map<std::string, std::string> kCrashDumpKeys = {
+    {"Process", "pid"},
+    {"Path", "path"},
+    {"Log Location", "crash_path"},
+    {"Identifier", "identifier"},
+    {"Version", "version"},
+    {"Parent Process", "parent"},
+    {"Responsible", "responsible"},
+    {"User ID", "user_id"},
+    {"Date/Time", "date_time"},
+    {"Crashed Thread", "crashed_thread"},
+    {"Exception Type", "exception_type"},
+    {"Exception Codes", "exception_codes"},
+    {"Exception Note", "exception_notes"},
+    {"rax", "rax"},
+    {"rbx", "rbx"},
+    {"rcx", "rcx"},
+    {"rdx", "rdx"},
+    {"rdi", "rdi"},
+    {"rsi", "rsi"},
+    {"rbp", "rbp"},
+    {"rsp", "rsp"},
+};
+
+void readCrashDump(const std::string& app_log, Row& r) {
+  r["crash_path"] = app_log;
+  std::string content;
+
+  if (!readFile(app_log, content).ok()) {
+    return;
+  }
+
+  // Variables for capturing the stack trace
+  boost::format crashed_thread_format("Thread %1% Crashed");
+  auto crashed_thread_seen = false;
+
+  auto lines = split(content, "\n");
+  for (auto it = lines.begin(); it != lines.end(); it++) {
+    auto line = *it;
+    // Tokenize first by colons
+    auto toks = split(line, ":");
+
+    if (toks.size() == 0 || kCrashDumpKeys.count(toks[0]) == 0) {
+      continue;
+    }
+
+    // Grab the most recent stack trace line of the crashed thread.
+    if (crashed_thread_seen && toks[0] == crashed_thread_format.str()) {
+      r["stack_trace"] = *(++it);
+      crashed_thread_seen = false;
+    }
+
+    // Process and grab all register values
+    if (toks[0] == "rax" || toks[0] == "rdi") {
+      toks = split(line, ": ");
+      for (size_t i = 0; i < toks.size(); i += 2) {
+        if ((i + 1) < toks.size()) // Ensure that we don't overflow
```

Comment:
1. nit, place the comment above or below the target line of code. Having comments after a line of code is weird.
2. Always use `{` and `}` to avoid careless `goto fail`s showing up by the next person editing the code.